webext/src/vimium.js

```diff
@@ -0,0 +1,1249 @@
+// The code in this file was copied and adapted from vimium's codebase.
```

Comment:
How did you get the raw code here? Was there a build process?

As I mentioned before, ideally I don't want to be committing this to the main code base. If there's a way to independently build it and include as a dependency, that would be ideal. Perhaps there's even a PR we could send to Vimium to help us?

Reply:
There are three main issues:
1) vimium's code is originally written in coffeescript. Assuming we don't want to switch over to using coffeescript and assuming we don't want a codebase that mixes javascript and coffeescript I copied the compiled code.
2) there doesn't exist a single file containing all relevant code (that we as browsh need) in vimium like the vimium.js file I created. The code is spread over multiple modules and not structured in a way that makes it easy to load in a 3rd party application. I think it's good to fetch only the relevant bits from vimium and not the 3/4ths we don't have much use for.
3) I made some (mostly minor) modifications to the code I documented at the top of vimium.js

I agree it's not the ideal solution, maybe we can create a issue on the vimium github site asking for the best way to share this particular code among the code bases. Unfortunately nobody answered my last question/issue I created there, so I went ahead anyway with the intention of making it work.

It should be possible to write a script(bash/python or the like) to recreate vimium.js. A lot of special case handling would be required (I took some liberty in splicing everything together) and we would have to maintain a patch set.

Depending on how much more functionality we want to use from vimium we could need to import additional code. To make the gi(focus first input) feature work there are some vimium bits we're still missing. Sometimes we can implement stuff our own way, I focused on importing vimium code for features that are most difficult to get right(i.e. link hinting).

Reply:
Oh you copied it from the inspector! Ha, ok nice. Well then what about just renaming `vimium.js` to `link_hinting.js`? It doesn't seem so necessary to have a formal dependency mechanism if we're just pulling out a quarter of the code. We should still of course reference the source of the code. And we can also mention at the top of the file that we would still ideally like to formalise the mechanism by which the code is included. Like you say, that would just be some scripts that convert the Coffee Script to JS at build time.

Reply:
@tombh 
I disagree about renaming it to link_hinting.js. It contains more than just the bits for link hinting, and I think vimium.js will grow a bit more and contain even more bits of vimium code, for example for input box selection. Link hinting is just an example of something not worth recreating the wheel for (lot's of work!). Adding more comments at the top is always a good idea.

Reply:
Sure, no problem. Ok, I'll have a quick go at importing with some build scripts. First thing though is to flesh out the tests so that any changes we make aren't going to break a bunch of the Vim features.

Reply:
@tombh 
It will be especially difficult to import the MiscVimium() function, because I created it. The code in there is for initializing stuff that other vimium code expects. It was put together from different vimium parts and also contains some modifications. This is the only part that you won't see in similar form in vimium.

Reply:
I think `MiscVimium` would be a good candidate for just including in the main Browsh code though. It's not very big. BTW I don't understand what any of that function does. It'd be good to have a few comments. That's another good reason for depending on the original Vimium Coffee Script, it already comes with some comments.

Reply:
@tombh 
The Non-MiscVimium-Vimium code was complaining about missing global objects and the like, so I had to figure out what those were and initialized them in MiscVimium. I would advise against putting it into main Browsh code, as it has no relation to any core Browsh logic, and would look very out of place. Good point about the comments not being included, maybe there's an option to compile CoffeeScript with intact comments. The missing comments didn't bother me much. Using coffee script for compiling our own vimium.js makes sense, but I do think the vimium.js should be precompiled in the repository and coffeescript shouldn't be required part of the toolchain for every Browsh developer.

Reply:
I created an issue at the github site of the vimium project about using their code in another application:
https://github.com/philc/vimium/issues/3133

So far there is no response.

Reply:
If I understand this already merged pull request correctly, this makes it possible to keep all types of comments in the compiled coffescript(which is javascript):
https://github.com/jashkenas/coffeescript/pull/4572